elfo-core/src/supervisor.rs

```diff
@@ -323,26 +332,33 @@ where
                 Err(panic) => ActorStatus::FAILED.with_details(panic_to_string(panic)),
             };
 
-            let should_restart = {
-                let object = sv.objects.get(&key).expect("where is the current actor?");
-                let actor = object.as_actor().expect("a supervisor stores only actors");
-
-                let rp_override = actor.restart_policy();
-                let restart_policy = rp_override.as_ref().unwrap_or(&sv.restart_policy);
-                let should_restart = match restart_policy.mode {
-                    RestartMode::Always => true,
-                    RestartMode::OnFailures => new_status.is_failed(),
-                    RestartMode::Never => false,
-                };
-
-                actor.set_status(new_status);
-                should_restart
-            };
-
-            let need_to_restart = should_restart && !sv.control.read().stop_spawning;
-            if need_to_restart {
-                let after = backoff.next();
-
+            let object = sv.objects.get(&key).expect("where is the current actor?");
+            let actor = object.as_actor().expect("a supervisor stores only actors");
+
+            let rp_override = actor.restart_policy();
+            // Select the restart policy with the next priority:
+            //     ctx.set_restart_policy(..) > config > blueprint (sv.restart_policy).
+            let restart_policy = rp_override
+                .as_ref()
+                .unwrap_or(config_restart_policy.as_ref().unwrap_or(&sv.restart_policy))
+                .clone();
+
+            let restarting_allowed =
+                restart_policy.restarting_allowed(&new_status) && !sv.control.read().stop_spawning;
+            actor.set_status(new_status);
+
+            drop(object);
+
+            let restart_after = restarting_allowed
+                .then(|| {
+                    let restart_params = restart_policy
+                        .restart_params()
+                        .expect("restart params are set if actor can restart");
+                    backoff.next(&restart_params)
```

Comment:
It means that until an actor dies, his policy cannot be changed. So, it's better to calculate the actual restart policy here, not at the start of `fn spawn`

Reply:
true, made adjustments